### PR TITLE
Bignum: Extract mod exp from prototype

### PR DESCRIPTION
**Pull request by yanesca**
## Description

The second PR of a series related to #6293 .

## Gatekeeper checklist

- [x] **changelog** not required
- [x] **backport** not required
- [x] **tests** provided

### CODE REVIEW

**Comment by yanesca** [comment]
Internal CI has passed and OpenCI fails on an unrelated issue.

**Comment by yanesca** [comment]
I am going to rework the test generation so that it will be easier to adapt when the improvements from #6577 and #6576 are merged.

That said, I would not like to wait for these PRs: these are not blockers and want to merge this PR independently of those two.

**Comment by yanesca** [comment]
Before that I would like to clean up the history a bit. Force pushing with the cleaner history and a minor fix.

**Comment by yanesca** [comment]
Internal CI has passed, Open CI reported and error, but there is no error on the corresponding CI job.

File: library/bignum_core.h [hunk 1]
```diff
@@ -496,6 +498,29 @@ int mbedtls_mpi_core_fill_random( mbedtls_mpi_uint *X, size_t X_limbs,
...
+ * \param E_limbs  The number of limbs in \p E.
+ * \param[in] RR   The precomputed residue of 2^{2*biL} modulo N, as a little
+ *                 endian array of length \p limbs.
+ *
+ * \return         \c 0 if successful.
+ * \return         #MBEDTLS_ERR_MPI_ALLOC_FAILED if a memory allocation failed.
+ */
+int mbedtls_mpi_core_exp_mod( mbedtls_mpi_uint *X,
+                              const mbedtls_mpi_uint *A,
+                              const mbedtls_mpi_uint *N, size_t limbs,
```
Comment (tom-cosgrove-arm) [inline library/bignum_core.h:520]: I really don't like having just `limbs` here. How does `AN_limbs` sound?

Also, and probably a nuisance, it might help if the implementation used the same parameter names

File: library/bignum_core.c [hunk 2]
```diff
@@ -582,6 +583,141 @@ int mbedtls_mpi_core_fill_random(
...
+    mbedtls_mpi_core_montmul( Wcur, Wcur, RR, n, N, n, mm, temp );
+    W1 = Wcur;
+    Wcur += n;
+    /* W[i+1] = W[i] * W[1], i >= 2 */
+    Wlast = W1;
+    for( size_t i=2; i < welem; i++, Wlast += n, Wcur += n )
+        mbedtls_mpi_core_montmul( Wcur, Wlast, W1, n, N, n, mm, temp );
+
+    /*
+     * Sliding window exponentiation
```
Comment (tom-cosgrove-arm) [inline library/bignum_core.c:660]: reminder: not sliding window any more

File: library/bignum_core.c [hunk 3]
```diff
@@ -582,6 +583,141 @@ int mbedtls_mpi_core_fill_random(
...
+    mbedtls_mpi_core_montmul( Wtbl, Wtbl, RR, n, N, n, mm, temp );
+    Wcur = Wtbl + n;
+    /* W[1] = A * R^2 * R^-1 mod N = A * R mod N */
+    memcpy( Wcur, A, n * ciL );
+    mbedtls_mpi_core_montmul( Wcur, Wcur, RR, n, N, n, mm, temp );
+    W1 = Wcur;
+    Wcur += n;
+    /* W[i+1] = W[i] * W[1], i >= 2 */
+    Wlast = W1;
+    for( size_t i=2; i < welem; i++, Wlast += n, Wcur += n )
```
Comment (tom-cosgrove-arm) [inline library/bignum_core.c:656]:
```suggestion
    for( size_t i = 2; i < welem; i++, Wlast += n, Wcur += n )
```

File: library/bignum_core.c [hunk 4]
```diff
@@ -582,6 +583,141 @@ int mbedtls_mpi_core_fill_random(
...
+#if( MBEDTLS_MPI_WINDOW_SIZE < 6 )
+    if( wsize > MBEDTLS_MPI_WINDOW_SIZE )
+        wsize = MBEDTLS_MPI_WINDOW_SIZE;
+#endif
+
+    return( wsize );
+}
+
+int mbedtls_mpi_core_exp_mod( mbedtls_mpi_uint *X,
+                              mbedtls_mpi_uint const *A,
```
Comment (tom-cosgrove-arm) [inline library/bignum_core.c:600]: consistency:
```suggestion
                              const mbedtls_mpi_uint *A,
```

File: library/bignum_core.c [hunk 5]
```diff
@@ -582,6 +583,141 @@ int mbedtls_mpi_core_fill_random(
...
+
+#if( MBEDTLS_MPI_WINDOW_SIZE < 6 )
+    if( wsize > MBEDTLS_MPI_WINDOW_SIZE )
+        wsize = MBEDTLS_MPI_WINDOW_SIZE;
+#endif
+
+    return( wsize );
+}
+
+int mbedtls_mpi_core_exp_mod( mbedtls_mpi_uint *X,
```
Comment (tom-cosgrove-arm) [inline library/bignum_core.c:599]: As per my comment in the `.h` file, would be great if this used the same variable names as we've agreed on (specifically `n` → `AN_limbs`, `E_len` → `E_limbs`)

File: library/bignum_core.c [hunk 6]
```diff
@@ -582,6 +583,141 @@ int mbedtls_mpi_core_fill_random(
...
+    Wtbl    = mempool;
+    Wselect = Wtbl    + table_limbs;
+    temp    = Wselect + wselect_limbs;
+
+    /*
+     * Window precomputation
+     */
+
+    /* W[0] = 1 (in Montgomery presentation) */
+    memset( Wtbl, 0, n * ciL ); Wtbl[0] = 1;
```
Comment (tom-cosgrove-arm) [inline library/bignum_core.c:646]: (should) one statement per line

File: scripts/mbedtls_dev/bignum_core.py [hunk 7]
```diff
@@ -818,6 +820,204 @@ def mpi_modmul_case_generate() -> None:
...
+# Ninth number generated by random.getrandbits(192) - seed(2,2), not a prime
+RANDOM_192_BIT_SEED_2_NO9 = "2a1be9cd8697bbd0e2520e33e44c50556c71c4a66148a86f"
+
+# Generated by bignum_common.gen_safe_prime(1024,3)
+SAFE_PRIME_1024_BIT_SEED_3 = ("c93ba7ec74d96f411ba008bdb78e63ff11bb5df46a51e16b"
+                              "2c9d156f8e4e18abf5e052cb01f47d0d1925a77f60991577"
+                              "e128fb6f52f34a27950a594baadd3d8057abeb222cf3cca9"
+                              "62db16abf79f2ada5bd29ab2f51244bf295eff9f6aaba130"
+                              "2efc449b128be75eeaca04bc3c1a155d11d14e8be32a2c82"
+                              "87b3996cf6ad5223")
```
Comment (tom-cosgrove-arm) [inline scripts/mbedtls_dev/bignum_core.py:942]: Since these have `1024` bits, do we need to worry about `MPI_MAX_BITS_LARGER_THAN_792`?

Comment (yanesca) [reply scripts/mbedtls_dev/bignum_core.py:942]: Functions for the old interface are using `MBEDTLS_MPI_MAX_SIZE` to determine the size of some stack allocated arrays. As long as we don't rely on `MBEDTLS_MPI_MAX_SIZE`, we don't need `MPI_MAX_BITS_LARGER_THAN_792`.

**Comment by yanesca** [comment]
@tom-cosgrove-arm I have applied the changes you suggested and the PR is ready for review again.

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM (modulo CI pass)

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Partial review: so far I've fully reviewed the library, not the tests.

I am satisfied that the code is functionally correct (no infinite loops, no buffer overflows, calculates what it should) and has a constant trace with respect to the exponent. However, I found the code very hard to follow. It doesn't group related parts, has many mutable variables with a broad scope, and the iteration logic is complex. In order to understand what was going on, I refactored the code and I found [the result](https://github.com/gilles-peskine-arm/mbedtls/tree/pr_6547-readability-20221116) a lot easier to follow. (And also a few bytes smaller in our reference Cortex-M0+ build.) I strongly recommend taking my refactored code.

In addition, `mempool` should be zeroized.

File: library/bignum_core.c [hunk 8]
```diff
@@ -582,6 +583,142 @@ int mbedtls_mpi_core_fill_random(
...
+    const size_t table_limbs   = welem * AN_limbs;
+    const size_t temp_limbs    = 2 * AN_limbs + 1;
+    const size_t wselect_limbs = AN_limbs;
+    const size_t total_limbs   = table_limbs + temp_limbs + wselect_limbs;
+
+    mempool = mbedtls_calloc( total_limbs, sizeof(mbedtls_mpi_uint) );
+    if( mempool == NULL )
+    {
+        ret = MBEDTLS_ERR_MPI_ALLOC_FAILED;
+        goto cleanup;
```
[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply library/bignum_core.c:634]: In a function where there is a "cleanup" label at the end, it's better to have all the variables declared at the top, to avoid the issue of potentially uninitialised variable access (when a goto happens before the variable is declared). If [you want some variables to have a reduced scope](https://github.com/Mbed-TLS/mbedtls/pull/6547/files#r1024378604), it's much better to have an early return and no "goto cleanup".

File: library/bignum_core.c [hunk 9]
```diff
@@ -582,6 +583,142 @@ int mbedtls_mpi_core_fill_random(
...
+    mbedtls_mpi_core_montmul( Wtbl, Wtbl, RR, AN_limbs, N, AN_limbs, mm, temp );
+    Wcur = Wtbl + AN_limbs;
+    /* W[1] = A * R^2 * R^-1 mod N = A * R mod N */
+    memcpy( Wcur, A, AN_limbs * ciL );
+    mbedtls_mpi_core_montmul( Wcur, Wcur, RR, AN_limbs, N, AN_limbs, mm, temp );
+    W1 = Wcur;
+    Wcur += AN_limbs;
+    /* W[i+1] = W[i] * W[1], i >= 2 */
+    Wlast = W1;
+    for( size_t i = 2; i < welem; i++, Wlast += AN_limbs, Wcur += AN_limbs )
```
[the remark this answers is not in the document]
Comment (tom-cosgrove-arm) [reply library/bignum_core.c:657]: Huh? It's pretty idiomatic in C to put increments like this within the for statement, especially when there's only one statement in the loop body. The `i++` is first, so it's clear that this is a simple counted for-loop, and having the updates to `Wlast` and `Wcur` here aids readability, IMO

File: library/bignum_core.c [hunk 10]
```diff
@@ -582,6 +582,141 @@ int mbedtls_mpi_core_fill_random(
 
 /* BEGIN MERGE SLOT 1 */
 
+static size_t mpi_exp_mod_get_window_size( size_t Ebits )
+{
+    size_t wsize = ( Ebits > 671 ) ? 6 : ( Ebits > 239 ) ? 5 :
```
[the remark this answers is not in the document]
Comment (yanesca) [reply library/bignum_core.c:588]: I am sorry, what do you mean by reference? A source where they came from? Or did you mean by adding an enum or named constant for them?

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/bignum_core.c:588]: I mean a citation (or a justification) for why we're using those particular thresholds to decide the window size.

**Comment by yanesca** [comment]
Rebased to remove commits adding the tests. (Doing rebase step-by-step in the hope that it will help review.)

**Comment by yanesca** [comment]
Rebased to current development head.

**Comment by yanesca** [comment]
Did the rebase in two steps: the first removes automated tests and should be easy to review and the second is a trivial rebase on top of current development head.

The first new commit is [07f2c69](https://github.com/Mbed-TLS/mbedtls/pull/6547/commits/07f2c69511303ddc667756c0bb56c90d737573b4).

File: library/bignum_core.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (yanesca) [reply library/bignum_core.c:588]: I don't think we have any good justification (at least not that I know of). These numbers go back to the beginning of our git history.

**Review by gilles-peskine-arm — COMMENTED** [review]
I'm happy with the library code as of 43d3de4977a86be1cf19bc6e3b7a558ab86737a6. I still haven't reviewed the tests.

**Comment by yanesca** [comment]
The CI run 3 days ago last time and it passed. The recent changes on development didn't impact this part of the code. Re-running the CI just to be sure.

The generated tests are up to date with the latest changes in the framework.

No rebase is necessary, the PR is ready for review as it is.

File: library/bignum_core.c [hunk 11]
```diff
@@ -582,6 +583,162 @@ int mbedtls_mpi_core_fill_random(
...
+     */
+
+    /* X = 1 (in Montgomery presentation) initially */
+    memcpy( X, Wtable, AN_limbs * ciL );
+
+    /* We'll process the bits of E from most significant
+     * (limb_index=E_limbs-1, E_bit_index=biL-1) to least significant
+     * (limb_index=0, E_bit_index=0). */
+    size_t E_limb_index = E_limbs;
+    size_t E_bit_index = 0;
```
Comment (tom-cosgrove-arm) [inline library/bignum_core.c:689]: (minor, not to change unless something else needs to) It's a bit weird saying we start from `(limb_index = E_limbs-1, E_bit_index = biL-1)` to then see

```
size_t E_limb_index = E_limbs;
size_t E_bit_index = 0;
```

which clearly does not match the comment and means one is immediately jumping forward to see what is missed.

I wonder about
```
/* E_bit_index == 0 is trigger to decrement E_limb_index and set E_bit_index = biL-1 */
size_t E_bit_index = 0;
```

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by tom-cosgrove-arm** [comment]
(ABI-API check false positive due to this PR being based on `development` prior to the CID changes going in)

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM
